cudamapper/src/overlapper.cpp

```diff
@@ -32,6 +32,28 @@ void Overlapper::filter_overlaps(std::vector<Overlap>& filtered_overlaps, const
                  valid_overlap);
 }
 
+void Overlapper::update_read_names(std::vector<Overlap>& overlaps,
+                                   const Index& index_query,
+                                   const Index& index_target)
+{
+#pragma omp parallel for
+    for (size_t i = 0; i < overlaps.size(); i++)
+    {
+        auto& o                      = overlaps[i];
+        std::string query_read_name  = index_query.read_id_to_read_name(o.query_read_id_);
+        std::string target_read_name = index_target.read_id_to_read_name(o.target_read_id_);
+
+        o.query_read_name_ = new char[query_read_name.length()];
+        strcpy(o.query_read_name_, query_read_name.c_str());
```

Comment:
Is copy to `query_read_name` actually necessary? Can't we simply do `strcpy(o.query_read_name_, index_query.read_id_to_read_name(o.query_read_id_).c_str());`
(make have `const std::string& query_read_name` if we want a separate variable for readability reasons)
@vellamike for this question

Reply:
yes, this is just the way it is for readability.

Reply:
Isn't this a write beyond bounds?
`query_read_name.length()` returns the number of chars in `query_read_name` without the null terminator, right?
`query_read_name.c_str()` returns a pointer to a c-string with a null terminator and `strcpy` copies the string until, but including, the null terminator. So it will write `query_read_name.length()+1` characters.

We should really get rid of handling raw arrays via pointers in the long run.

Reply:
Hmm, yes, that looks bad.

Regarding raw pointers vs string, I fully agree. The problem is that Thrust had problems working with structs that have string in them so that's why we decided to stick with raw pointers.
Mike, Kamesh and I all had a quick look at it and couldn't find a solution, but we should invest some more time in it.

Reply:
If we don't require null-terminated c-strings, we could just not use `strcpy` here and do `std::copy_n(query_read_name.data(), query_read_name.size(), o.query_read_name_);` instead.
Note that writing strings, e.g. to stdout, might require them to be null terminated though.

Reply:
Apparently we use `printf` for printing out the results (https://github.com/clara-genomics/ClaraGenomicsAnalysis/blob/bb77507ec301e03b48539d336bcdc6eb4ce9c078/cudamapper/src/overlapper.cpp#L40), but that still requires the null-terminated string

Reply:
I just tried switching to `const std::string& query_read_name`, but there was no performance difference. I assume compiler was able to avoid that copy to `std::string query_read_name` as `index_query` is `const`.
I'd say we should anyway switch to `const std::string& query_read_name` and not rely on compiler to do it

Reply:
changed it to `const std::string&` and also to be safe, increased the size by one char `new char[query_read_name.length()+1]`.
Also should note in addition to possible compiler optimizations, there are 2 additional reasons in this PR not to see any performance difference by changing to  `const std::string&`: 1- it runs async 2- it runs on a rather small set of filtered overlaps, rather than looping on large arrays as in previous versions